src/main/scala/edu/ie3/simona/agent/participant/hp/HpAgentFundamentals.scala

```diff
@@ -0,0 +1,386 @@
+      inputModel: InputModelContainer[HpInput],
+      modelConfig: HpRuntimeConfig,
+      services: Option[Vector[SecondaryDataService[_ <: Data.SecondaryData]]],
+      simulationStartDate: ZonedDateTime,
+      simulationEndDate: ZonedDateTime,
+      resolution: Long,
+      requestVoltageDeviationThreshold: Double,
+      outputConfig: NotifierConfig
+  ): BaseStateData.ParticipantModelBaseStateData[
+    ApparentPowerAndHeat,
+    HpRelevantData,
+    HpModel
+  ] = {
+    if (!services.exists(_.map(_.getClass).containsSlice(neededServices)))
+      throw new AgentInitializationException(
+        "HpAgent cannot be initialized without its needed services."
+      )
+
+    inputModel match {
+      case withHeatContainer: WithHeatInputContainer[HpInput] =>
+        /* Build the calculation model */
+        val model = buildModel(
+          withHeatContainer,
+          modelConfig,
+          simulationStartDate,
+          simulationEndDate
+        )
+        ParticipantModelBaseStateData[
+          ApparentPowerAndHeat,
+          HpRelevantData,
+          HpModel
+        ](
+          simulationStartDate,
+          simulationEndDate,
+          model,
+          services,
+          outputConfig,
+          SortedSet.empty,
+          Map.empty,
+          requestVoltageDeviationThreshold,
+          ValueStore.forVoltage(
+            resolution,
+            Each(
+              inputModel.electricalInputModel.getNode
+                .getvTarget()
+                .to(PU)
+                .getValue
+                .doubleValue()
+            )
+          ),
+          ValueStore(resolution),
+          ValueStore(resolution),
+          ValueStore(resolution)
+        )
+      case unsupported =>
+        throw new AgentInitializationException(
+          s"HpAgent cannot be initialized with wrong init state data of type '${unsupported.getClass.getName}'. " +
+            s"Expected: '${WithHeatInputContainer.getClass.getName}[${classOf[HpInput].getName}]'."
+        )
+    }
+  }
+
+  /** Abstract method to build the calculation model from input
+    *
+    * @param inputModel
+    *   Input model description
+    * @param modelConfig
+    *   Configuration for the model
+    * @param simulationStartDate
+    *   Wall clock time of first instant in simulation
+    * @param simulationEndDate
+    *   Wall clock time of last instant in simulation
+    * @return
+    */
+  override def buildModel(
+      inputModel: InputModelContainer[HpInput],
+      modelConfig: HpRuntimeConfig,
+      simulationStartDate: ZonedDateTime,
+      simulationEndDate: ZonedDateTime
+  ): HpModel = {
+    val thermalHouseInput = {
+      inputModel match {
+        case ParticipantStateData.SimpleInputContainer(_) =>
+          None
+        case WithHeatInputContainer(_, thermalGrid) =>
+          /* Build the thermal model */
+          thermalGrid.houses().stream().findFirst().asScala
+      }
+    }.getOrElse {
+      new ThermalHouseInput(
```

Comment:
Do we want to create some random thermal bus if the correct one can't be found? From my point of view an exception would be better here. Or?

Reply:
After discussing this shortly we opt for an exception in this case.